### PR TITLE
Fixed nested themes not being republished on outer theme changes

### PR DESCRIPTION
**Pull request by Andarist**
**What**:

This fixes a bug when having nested `ThemeProvider`s and updating outer theme the change is not reaching the `StyledComponent`
```jsx
<ThemeProvider theme={{}}>
	<ThemeProvider theme={{}}>
		<StyledComponent>
	</ThemeProvider>
</ThemeProvider>
```

**Why**:

This is a bug fix.

**How**:
Using publish in the nested `ThemeProvider`s subscriptions.

**Checklist**:
- [x] Tests
- [x] Code complete

Repro provided by @mitchellhamilton - https://codesandbox.io/s/6j42qx5z23

### CODE REVIEW

File: src/theme-provider.js [hunk 1]
```diff
@@ -53,13 +58,12 @@ class ThemeProvider extends React.Component {
     // set broadcast state by merging outer theme with own
     if (this.context[CHANNEL]) {
       this.setOuterTheme(this.context[CHANNEL].getState())
-      this.broadcast.setState(this.getTheme())
```
Comment (kentcdodds) [inline src/theme-provider.js]: Can you explain why this was removed? It seems unrelated.

Comment (Andarist) [reply src/theme-provider.js]: publishTheme method is doing a setState and setOuterTheme uses now publishTheme under the hood

**Comment by Andarist** [comment]
@kentcdodds do u have any other reviewer in mind? The chane is pretty straightforward and the codebase is tested so it shouldnt break anything ;)

File: src/theme-provider.js [hunk 2]
```diff
@@ -53,13 +58,12 @@ class ThemeProvider extends React.Component {
...
     }
   }
 
   componentWillReceiveProps(nextProps) {
     if (this.props.theme !== nextProps.theme) {
-      this.broadcast.setState(this.getTheme(nextProps.theme))
+      this.publishTheme(nextProps.theme)
```
Comment (paulmolluzzo) [inline src/theme-provider.js]: Pardon my ignorance, but how has this made an effectual change? `publishTheme` is calling `broadcast.setState(this.getTheme(nextProps.theme))` which is was was here before. It's only been moved to a separate method.

File: src/theme-provider.js (hunk 1 above)
Comment (paulmolluzzo) [reply src/theme-provider.js]: I think this is the real fix, and `publishTheme` is only incidentally changed.

This line was setting the state with no argument, which defaults to the `props.theme` here https://github.com/paypal/glamorous/blob/cb6de3abe6b3a7e2643828f4239e0487d4f80455/src/theme-provider.js#L21, meaning it's not setting the theme to the *next* theme (it's just resetting it to the theme it already has).

A lot of the theming work was done by @vesparny, so if I'm wrong hopefully he can correct me. 😄

File: src/theme-provider.js (hunk 2 above)
Comment (Andarist) [reply src/theme-provider.js]: oh, this part is actually not a fix of any sorts, i just have extracted it to separate method so it would be easier to reuse it in `setOuterTheme`

Comment (paulmolluzzo) [reply src/theme-provider.js]: OK, thanks for clarifying.

File: src/theme-provider.js (hunk 1 above)
Comment (Andarist) [reply src/theme-provider.js]: Yeah, this part of defaulting to `this.props.theme` is a making a control flow a little bit iffy but I left this part untouched. Personally I would prefer explicit arguments, so that `publishTheme` would always have to receive `getTheme`'s result as argument which in turn would always be called with `this.props.theme` if necessary.

Also I've noticed a different small issue too - broadcast is initialized as:
```js
broadcast = brcast(this.props.theme)
```
but that gives it untrue initialState in case of nested ThemeProvider. I think creation of the `broadcast` should be delayed.

Comment (paulmolluzzo) [reply src/theme-provider.js]: > but that gives it untrue initialState in case of nested ThemeProvider. I think creation of the broadcast should be delayed.

I'm pretty sure I follow, but to be clear, are you suggesting moving it to something like `componentWillMount`? With this PR, the theme that was (possibly) merged with the outer theme is now going to be set properly inside `componentWillMount()`, so that untrue initial state won't really ever be in effect.

Did I follow your point correctly? If not any clarification would be great. 🙌

Comment (Andarist) [reply src/theme-provider.js]: Yeah, that was exactly my point :)

Comment (paulmolluzzo) [reply src/theme-provider.js]: Cool, thanks! It makes sense to me. Maybe you want to do that as a follow-up PR?

Comment (Andarist) [reply src/theme-provider.js]: Sure thing, would prefer to create it once we settle on this one to avoid potential conflicts

Comment (kentcdodds) [reply src/theme-provider.js]: Thanks for the discussion here!
